lib/src/command/token_add.dart

```diff
@@ -61,4 +63,33 @@ class TokenAddCommand extends PubCommand {
       throw ApplicationException('Token is not provided within 15 minutes.');
     }
   }
+
+  Future<void> _addTokenFromStdin(Uri hostedUrl) async {
+    final token = await stdinPrompt('Enter secret token:')
+        .timeout(const Duration(minutes: 15));
+    if (token.isEmpty) {
+      usageException('Token is not provided.');
+    }
+
+    tokenStore.addCredential(Credential.token(hostedUrl, token));
+    log.message(
+      'Requests to $hostedUrl will now be authenticated using the secret '
+      'token.',
+    );
+  }
+
+  Future<void> _addTokenFromEnv(Uri hostedUrl) async {
+    tokenStore.addCredential(Credential.env(hostedUrl, envVar));
+    log.message(
+      'Requests to $hostedUrl will now be authenticated using the secret '
+      'token stored in environment variable.',
+    );
+
+    if (!Platform.environment.containsKey(envVar)) {
+      // If environment variable doesn't exist when
+      // pub token add <hosted-url> --env-var <ENV_VAR> is called, we should
+      // print a warning.
+      log.warning('Environment variable $envVar is not defined.');
```

Comment:
I'm not sure we should warn here. There is no reason the variable should be defined already - instead it should be defined when using.

@jonasfj do you agree?

Reply:
I think we should warn... but I can see the argument against it given how github actions work (ie. you only have to pass the env var with the secret to the command that does the operation).

For example, the following would produce a warning:
```yaml
jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/dart...@v...
      - run: dart pub token add pub.dev --env=AUTH_TOKEN
      - run: dart pub publish
        env:
          AUTH_TOKEN: ${{ secrets.PUB_TOKEN }}
```

The warning probably wouldn't cause much alarm if the following command succeeds :D

on the other hand, this wouldn't:
```yaml
jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/dart...@v...
      - run: dart pub token add pub.dev --env=AUTH_TOKEN
        env:
          AUTH_TOKEN: ${{ secrets.PUB_TOKEN }}
      - run: dart pub publish
        env:
          AUTH_TOKEN: ${{ secrets.PUB_TOKEN }}
```

Doing option two here is more work, but it also wouldn't be less secure.

-----------

I think that to help users do the right thing we need to do either:
 * (A) restrict allowed env var names to `[a-zA-Z_][a-zA-Z0-9_]*`, and/or,
 * (B) warn if the environment variable isn't defined when `dart token add <hosted-url> --env-var=<NAME>` is called.

IMO, we should lean towards keeping things simple and do both (A) and (B).
But I could also be sold on only doing one of these two things.

IMO; we can always remove the warning in (B), or re-introduce it later.

We can easily remove a restriction like (A), but we can't add it later.

Hence, I suggest we do (A), and (B). I know sigurdm@ leans a little towards doing neither, but I'm happy letting him pick at-least one of them that we should do.

Note. I think it's important that we do either (A) or (B) to help users when they are making typos.